mlflow/store/sqlalchemy_store.py

```diff
@@ -184,6 +184,7 @@ def create_experiment(self, name, artifact_location=None):
                 raise MlflowException('Experiment(name={}) already exists. '
                                       'Error: {}'.format(name, str(e)), RESOURCE_ALREADY_EXISTS)
 
+            session.flush()
```

Comment:
This `session.flush()` call is necessary to actually insert the experiment into the DB and give it an ID - otherwise  `experiment.experiment_id` is incorrectly set to `None` below

Reply:
Isn't the session automatically committed after exiting the managed session context? The commit operation always performs a flush. Any idea why this doesn't work without explicitly flushing?

Reply:
Good question, it's because the `__exit__` method of the managed session context object doesn't run till after the return value is evaluated. This is why logic like

```
$ echo "blahblah" > /tmp/my-file
$ python
Python 2.7.15 |Anaconda, Inc.| (default, May  1 2018, 18:37:05) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def my_method():
...   with open("/tmp/my-file") as f:
...     return f.readlines()
>>> my_method()
['blahblah\n']
``` 

works (the file is only closed by `__exit__` after the return value is computed). It's kinda hard to find exact docs describing this behavior but it's alluded to [here](https://www.python.org/dev/peps/pep-0343/#id38):

> "The calling convention for mgr.__exit__() is as follows. If the finally-suite was reached through normal completion of BLOCK or through a non-local goto (a break, continue or return statement in BLOCK), mgr.__exit__() is called with three None arguments"

Reply:
Awesome, thanks for the detailed explanation!

Reply:
Can we update `test_create_experiments` to validate that the returned `experiment_id` is not `None`?